File: 39.combination-sum.rb
#
# @lc app=leetcode id=39 lang=ruby
#
# [39] Combination Sum
#

# @lc code=start
# @param {Integer[]} candidates
# @param {Integer} target
# @return {Integer[][]}
# def combination_sum(candidates, target)
#   dp = [[[]]]
#
#   (1..target).each do |i|
#     candidates.each do |candidate|
#       if candidate <= i && !dp[i - candidate].nil?
#         dp[i] ||= []
#         dp[i - candidate].each do |v|
#           dp[i].push(v.clone.append(candidate).sort)
#         end
#         dp[i] = [[candidate]] if dp[i] == []
#         dp[i] = dp[i].uniq
#       end
#     end
#   end
#
#   dp[target] || []
# end
def combination_sum(candidates, target)
  dp = Array.new(target + 1) { [] }

  candidates.each do |candidate|
    (candidate..target).each do |i|
      dp[i].push([candidate]) if candidate == i

      dp[i - candidate].each do |v|
        dp[i].push(v + [candidate])
      end
    end
  end

  dp[target]
end
# @lc code=end

File: 39.combination-sum.go
/*
 * @lc app=leetcode id=39 lang=golang
 *
 * [39] Combination Sum
 */

package main

// @lc code=start
func combinationSum(candidates []int, target int) [][]int {
	dp := make([][][]int, target+1)
	for i := range dp {
		dp[i] = [][]int{}
	}

	for i := 1; i <= target; i++ {
		for _, candidate := range candidates {
			if candidate <= i {
				dp[i] = append(dp[i], dp[i-candidate]...)
				for j := 0; j < len(dp[i]); j++ {
					dp[i][j] = append(dp[i][j], candidate)
				}

			}
		}
	}

	return dp[target]
}

// @lc code=end
